makefile

```diff
@@ -587,51 +587,52 @@ distclean: clean config/clean doc/clean
-	$(INSTALL) -m 640 $(SRC)config/tconfig.hpp $(PREFIX_INC)/mfem/config
+	$(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/config
+	$(INSTALL) -m 644 $(BLD)config/_config.hpp $(PREFIX_INC)/mfem/config
+	$(INSTALL) -m 644 $(SRC)config/config.hpp $(PREFIX_INC)/mfem/config
+	$(INSTALL) -m 644 $(SRC)config/tconfig.hpp $(PREFIX_INC)/mfem/config
 # install remaining includes in each subdirectory
 	for dir in $(DIRS); do \
-	   mkdir -p $(PREFIX_INC)/mfem/$$dir && \
-	   $(INSTALL) -m 640 $(SRC)$$dir/*.hpp $(PREFIX_INC)/mfem/$$dir; \
+	   $(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/$$dir && \
+	   $(INSTALL) -m 644 $(SRC)$$dir/*.hpp $(PREFIX_INC)/mfem/$$dir; \
 	done
 # install *.okl files
 	for dir in $(OKL_DIRS); do \
-	   mkdir -p $(PREFIX_INC)/mfem/$$dir && \
-	   $(INSTALL) -m 640 $(SRC)$$dir/*.okl $(PREFIX_INC)/mfem/$$dir; \
+	   $(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/$$dir && \
+	   $(INSTALL) -m 644 $(SRC)$$dir/*.okl $(PREFIX_INC)/mfem/$$dir; \
 	done
 # install libCEED q-function headers
-	mkdir -p $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
-	$(INSTALL) -m 640 $(SRC)fem/ceed/integrators/mass/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
-	mkdir -p $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
-	$(INSTALL) -m 640 $(SRC)fem/ceed/integrators/convection/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
-	mkdir -p $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
-	$(INSTALL) -m 640 $(SRC)fem/ceed/integrators/diffusion/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
-	mkdir -p $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
-	$(INSTALL) -m 640 $(SRC)fem/ceed/integrators/nlconvection/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
+	$(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
+	$(INSTALL) -m 644 $(SRC)fem/ceed/integrators/mass/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
+	$(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
+	$(INSTALL) -m 644 $(SRC)fem/ceed/integrators/convection/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
+	$(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
+	$(INSTALL) -m 644 $(SRC)fem/ceed/integrators/diffusion/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
+	$(INSTALL) -m 755 -d $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
+	$(INSTALL) -m 644 $(SRC)fem/ceed/integrators/nlconvection/*.h $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
 # install config.mk in $(PREFIX_SHARE)
-	mkdir -p $(PREFIX_SHARE)
+	$(INSTALL) -m 755 -d $(PREFIX_SHARE)
 	$(MAKE) -C $(BLD)config config-mk CONFIG_MK=config-install.mk
-	$(INSTALL) -m 640 $(BLD)config/config-install.mk $(PREFIX_SHARE)/config.mk
+	$(INSTALL) -m 644 $(BLD)config/config-install.mk $(PREFIX_SHARE)/config.mk
 	rm -f $(BLD)config/config-install.mk
 # install test.mk in $(PREFIX_SHARE)
-	$(INSTALL) -m 640 $(SRC)config/test.mk $(PREFIX_SHARE)/test.mk
+	$(INSTALL) -m 644 $(SRC)config/test.mk $(PREFIX_SHARE)/test.mk
```

Comment:
Hi @nmnobre,

Here's what I suggest we use with regard to install permissions -- it is very similar to what you have; compared to the current `master`, it uses the same permission for "others" as for "group"; it also allows easy switch to other permissions by setting the make variables `INSTALL_DEF_PERM`, `INSTALL_BIN_PERM`, and `INSTALL_DIR_MASK` on the command line or in `user.mk`

```suggestion
# Install permissions for everything except directories and binaries:
INSTALL_DEF_PERM ?= 644
# Install permissions for binaries:
INSTALL_BIN_PERM ?= 755
# Install umask for creating directories (and symlinks on MacOS/BSD).
# We use 'umask' because 'mkdir -p' (and 'install -d') do not use the mode
# specified with the '-m' flag when creating non-existent parent directories.
INSTALL_DIR_MASK ?= 022

INSTALL_SHARED_LIB = $(MFEM_CXX) $(MFEM_LINK_FLAGS) $(INSTALL_SOFLAGS)\
   $(OBJECT_FILES) $(EXT_LIBS) -o $(PREFIX_LIB)/libmfem.$(SO_VER) && \
   cd $(PREFIX_LIB) && chmod $(INSTALL_BIN_PERM) libmfem.$(SO_VER) && \
   ( umask $(INSTALL_DIR_MASK) && ln -sf libmfem.$(SO_VER) libmfem.$(SO_EXT) )

# Shortcuts, not to be modified on command line.
# WARNING: $(MKINSTALLDIR) changes the umask for commands following it as part
#          of the same shell expression unless it is placed inside '()' to be
#          executed in a sub-shell.
override MKINSTALLDIR = umask $(INSTALL_DIR_MASK) && mkdir -p
override INSTALLDEF   = $(INSTALL) -m $(INSTALL_DEF_PERM)

install: $(if $(static),$(BLD)libmfem.a) $(if $(shared),$(BLD)libmfem.$(SO_EXT))
	$(MKINSTALLDIR) $(PREFIX_LIB)
# install static and/or shared library
	$(if $(static),$(INSTALLDEF) $(BLD)libmfem.a $(PREFIX_LIB))
	$(if $(shared),$(INSTALL_SHARED_LIB))
# install top level includes
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem
	$(INSTALLDEF) $(SRC)mfem.hpp $(SRC)mfem-performance.hpp \
	   $(PREFIX_INC)/mfem
	for hdr in mfem.hpp mfem-performance.hpp; do \
	   printf '// Auto-generated file.\n#include "mfem/'$$hdr'"\n' \
	      > $(PREFIX_INC)/$$hdr && \
	   chmod $(INSTALL_DEF_PERM) $(PREFIX_INC)/$$hdr; done
# install config include
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/config
	$(INSTALLDEF) $(BLD)config/_config.hpp $(PREFIX_INC)/mfem/config
	$(INSTALLDEF) $(SRC)config/config.hpp $(PREFIX_INC)/mfem/config
	$(INSTALLDEF) $(SRC)config/tconfig.hpp $(PREFIX_INC)/mfem/config
# install remaining includes in each subdirectory
	for dir in $(DIRS); do \
	   ( $(MKINSTALLDIR) $(PREFIX_INC)/mfem/$$dir ) && \
	   $(INSTALLDEF) $(SRC)$$dir/*.hpp $(PREFIX_INC)/mfem/$$dir; \
	done
# install *.okl files
	for dir in $(OKL_DIRS); do \
	   ( $(MKINSTALLDIR) $(PREFIX_INC)/mfem/$$dir ) && \
	   $(INSTALLDEF) $(SRC)$$dir/*.okl $(PREFIX_INC)/mfem/$$dir; \
	done
# install libCEED q-function headers
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/mass/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/mass
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/convection/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/convection
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/diffusion/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/diffusion
	$(MKINSTALLDIR) $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
	$(INSTALLDEF) $(SRC)fem/ceed/integrators/nlconvection/*.h \
	   $(PREFIX_INC)/mfem/fem/ceed/integrators/nlconvection
# install config.mk in $(PREFIX_SHARE)
	$(MKINSTALLDIR) $(PREFIX_SHARE)
	$(MAKE) -C $(BLD)config config-mk CONFIG_MK=config-install.mk
	$(INSTALLDEF) $(BLD)config/config-install.mk $(PREFIX_SHARE)/config.mk
	rm -f $(BLD)config/config-install.mk
# install test.mk in $(PREFIX_SHARE)
	$(INSTALLDEF) $(SRC)config/test.mk $(PREFIX_SHARE)
```

Reply:
I like your idea, so here's what I suggest (and just committed):

- We roll back to `mkdir -p` to create the necessary directories. This sets the same permissions for parent directories and honours the user's umask. This was never a problem, so perhaps I should've never touched it. My bad.
- We use `cp` to copy libraries, headers and makefiles to the new directories. This will always subtract the user's umask from the source files permissions. Then, we just do `chmod +x` for the shared library file.

I think everybody will be happy with this solution:
- Typical users, like me, who probably never touched their umask, so it's still at 022, will get 755 for all directories and 644 for most files (except the shared object). This is what I expect will happen most often.
- People at gov labs might've different umask settings (maybe 026), either set by themselves or by some system administrator, and these will still be honoured with the suggested solution. Most files will then be installed with 640 as before, and directories with 751.
- It avoids the extra makefile variables, the user can simply tweak their umask to get the desired permissions.
 
How does that sound?

Reply:
I don't think we want to switch from `$(INSTALL)` to `cp` -- I know that users use the `INSTALL` variable to adjust things; and we don't want to create confusion by setting `INSTALL` to `cp`. Also, having the explicit variables for permissions makes it more obvious for users how to adjust those without having to figure out how they can have some control by modifying their `umask`.

Also, I have a tweak in the shared library installation command that I want to include (it affects MacOS).

Basically, I'll re-iterate my suggestion to apply the changes I posted above.

Reply:
Let me know what you think of this. Instead of setting `INSTALL_DIR_MASK`, the user may set `INSTALL_DIR_PERM` which I think might be less confusing and follows the same format of the other two new variables. Otherwise, it's as you suggested. With your permission, I'd like to squash some commits and force push.

Reply:
> Let me know what you think of this.

Looks good to me. Thanks!

> Instead of setting `INSTALL_DIR_MASK`, the user may set `INSTALL_DIR_PERM` which I think might be less confusing and follows the same format of the other two new variables.

Yes, this is a good change.

> Otherwise, it's as you suggested. If you're happy with this, and with your permission, I'd like to squash some commits and force push.

We can also just squash-merge the PR. If, however, you prefer to have a sequence of commits from the PR, I don't mind if you want to force-push. Just make sure the editor, @tzanio, and the other reviewers, @justinlaughlin, are also fine with that.

Reply:
> We can also just squash-merge the PR.

Indeed, I'm happy with that too.

> If, however, you prefer to have a sequence of commits from the PR

It just seemed like that's what you usually do judging from recent commit history, that's all.
I'll leave it to @tzanio to decide. I'm happy either way really.

Reply:
> I'd like to squash some commits and force push.

Please go ahead and do that